File: python/wax/complex_operations/recurrent_transfer.py
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Final, Iterable, cast

from wax._private.base_api import WaxBaseApi
from wax._private.operation_base import ConvertedToProtoOperation, OperationBase
from wax._private.proto.recurrent_transfer_extension_pb2 import recurrent_transfer_extension, recurrent_transfer_pair_id
from wax.exceptions.asset_errors import UnexpectedAssetTypeError
from wax.models.asset import AssetName, NaiAsset
from wax.proto.operations import recurrent_transfer

if TYPE_CHECKING:
    from wax import IWaxBaseInterface
    from wax.models.basic import AccountName

DEFAULT_EXECUTIONS: Final[int] = 2
DEFAULT_RECURRENCE: Final[int] = 24


@dataclass
class RecurrentTransferData:
    """
    Data needed to define recurrent transfer operation.

    Attributes:
        from_account: Account to transfer asset from.
        to_account: Account to transfer asset to. Cannot set a transfer to yourself.
        amount: The amount of asset to transfer. Allowed assets: **HIVE** and **HBD**.
                 In case of removal - amount is not needed.
        pair_id: Since HF 28, if user has more than one recurrent transfer to the same receiver
                 or creates the recurrent transfer using `pair_id`, they must specify it in order
                 to update or remove the defined recurrent transfer.
        executions: How many times the recurrent payment will be executed.
                 Executions must be at least 2; if set to 1, the transfer will not execute.
        recurrence: How often the payment is triggered, in hours. First transfer is immediate.
                 Minimum value is 24 h.
        memo: Optional memo for the transfer. Must be shorter than 2048 characters.
    """

    from_account: AccountName
    to_account: AccountName
    amount: NaiAsset | None = None
    pair_id: int | None = None
    executions: int = DEFAULT_EXECUTIONS
    recurrence: int = DEFAULT_RECURRENCE
    memo: str = ""


class RecurrentTransferOperationBase(OperationBase):
    """Base class for the recurrent transfer define and removal classes."""

    def __init__(self, data: RecurrentTransferData) -> None:
        super().__init__()
        self.data = data
        self.recurrent_transfer = self.get_recurrent_transfer_proto_operation() if data.amount is not None else None

    def finalize(self, api: IWaxBaseInterface) -> Iterable[ConvertedToProtoOperation]:
        if not cast(WaxBaseApi, api).check_is_proper_asset(
            [AssetName.Hive, AssetName.Hbd], self.recurrent_transfer.amount
        ):
            raise UnexpectedAssetTypeError(self.recurrent_transfer.amount, [AssetName.Hive, AssetName.Hbd])

        return [self.recurrent_transfer]

    def get_recurrent_transfer_proto_operation(self) -> recurrent_transfer:
        """
        Creates recurrent_transfer operation.

        Returns:
            A recurrent transfer operation.
        """
        return recurrent_transfer(
            from_account=self.data.from_account,
            to_account=self.data.to_account,
            amount=self.data.amount,
            extensions=[self._get_default_extensions(self.data.pair_id)] if self.data.pair_id is not None else [],
            executions=self.data.executions,
            recurrence=self.data.recurrence,
            memo=self.data.memo,
        )

    def _get_default_extensions(self, pair_id: int) -> recurrent_transfer_extension:
        """
        Creates a recurrent transfer extension with the specified pair ID.

        Args:
            pair_id: The unique identifier for the recurrent transfer pair.

        Raises:
            ValueError: If the provided pair_id is less than or equal to zero.

        Returns:
            recurrent_transfer_extension: The protobuf extension with the pair ID set.
        """
        if pair_id <= 0:
            raise ValueError("Pair id must be greater than 0")

        return recurrent_transfer_extension(recurrent_transfer_pair_id=recurrent_transfer_pair_id(pair_id=pair_id))


class DefineRecurrentTransferOperation(RecurrentTransferOperationBase):
    """
    Operation class for defining a recurrent transfer.

    This class validates that the amount is greater than zero and
    provides a name identifier for the operation.

    - The `amount` must be strictly greater than zero. A `ValueError` is raised if `amount` is None or 0.
    - If a recurrent transfer between the same `from` and `to` accounts already exists:
        - If the `recurrence` value is unchanged, the next execution will follow the original schedule.
        - If the `recurrence` value is changed, the next execution will be scheduled for: `update date + recurrence`.
          In this case, **no** transfer will be executed on the update date itself.
    - Since HF28, users may define multiple recurrent transfers between the same accounts by using `pair_id`
      in the extensions.
    - A single account can define up to **255** recurrent transfers to other accounts.
    - The final execution date of the recurrent transfer must be **no more than 730 days** in the future.

    Raises:
        ValueError: If the transfer amount is None or less than or equal to 0.
    """

    def __init__(self, data: RecurrentTransferData) -> None:
        amount = data.amount

        if amount is None:
            raise ValueError("Amount of the recurrent transfer must be provided.")

        if int(amount.amount) <= 0:
            raise ValueError("Amount of the recurrent transfer must be greater than 0.")
        super().__init__(data)


class RecurrentTransferRemovalOperation(RecurrentTransferOperationBase):
    """
    Operation class for removing a recurrent transfer.

    Ensures that amount is set to zero if not provided or invalid.

    - If multiple recurrent transfers exist between the same `from` and `to` accounts,
      a `pair_id` must be specified in the extensions to identify the exact one to be removed.
    - When this operation is executed, it **does not** generate a `fill_recurrent_transfer_operation` virtual operation.
    """

    def __init__(self, from_account: AccountName, to_account: AccountName, pair_id: int | None = None) -> None:
        super().__init__(RecurrentTransferData(from_account, to_account, pair_id=pair_id))

    def finalize(self, api: IWaxBaseInterface) -> Iterable[ConvertedToProtoOperation]:
        self.data.amount = api.hive.satoshis(0)

        self.recurrent_transfer = self.get_recurrent_transfer_proto_operation()
        return super().finalize(api)

File: rust/wax/src/complex_operations/recurrent_transfer.rs
use wax_core::proto;

use crate::WaxError;
use crate::foundation::WaxFoundation;
use crate::interfaces::OperationBuilder;
use crate::models::asset::{AssetName, NaiAssetConvertible};
use crate::models::basic::AccountName;

const DEFAULT_RECURRENCE_HOURS: u32 = 24;
const DEFAULT_EXECUTIONS: u32 = 2;

/// Coerces `amount` to HIVE or HBD, mirroring TS's multi-symbol
/// `createAssetWithRequiredSymbol([HBD, HIVE], …)`.
fn coerce_hive_or_hbd(
    foundation: &dyn WaxFoundation,
    amount: NaiAssetConvertible,
) -> Result<proto::Asset, WaxError> {
    // Try HIVE first to match the dominant case; fall through to HBD if the
    // caller supplied HBD instead. Either symbol is valid for recurrent
    // transfers per the protocol spec.
    let try_hive =
        foundation.create_asset_with_required_symbol(AssetName::Hive, amount.clone());
    match try_hive {
        Ok(asset) => Ok(asset),
        Err(_) => foundation.create_asset_with_required_symbol(AssetName::Hbd, amount),
    }
}

fn build_extensions(pair_id: Option<u32>) -> Vec<proto::RecurrentTransferExtension> {
    match pair_id {
        Some(pair_id) if pair_id != 0 => vec![proto::RecurrentTransferExtension {
            value: Some(
                proto::recurrent_transfer_extension::Value::RecurrentTransferPairId(
                    proto::RecurrentTransferPairId { pair_id },
                ),
            ),
        }],
        _ => Vec::new(),
    }
}

/// Builder for the `define` / `update` flavour of recurrent transfer, mirroring
/// `DefineRecurrentTransferOperation` from
/// `ts/wasm/lib/detailed/complex_operations/recurrent_transfer.ts`.
#[derive(Debug, Clone)]
pub struct DefineRecurrentTransferOperation {
    pub from_account: AccountName,
    pub to_account: AccountName,
    pub amount: NaiAssetConvertible,
    pub memo: Option<String>,
    pub recurrence: Option<u32>,
    pub executions: Option<u32>,
    pub pair_id: Option<u32>,
}

impl OperationBuilder for DefineRecurrentTransferOperation {
    fn finalize(
        self,
        foundation: &dyn WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        let amount = coerce_hive_or_hbd(foundation, self.amount)?;
        if amount.amount == "0" {
            return Err(WaxError::new("Amount must be greater than 0"));
        }

        Ok(vec![proto::Operation {
            value: Some(proto::operation::Value::RecurrentTransferOperation(
                proto::RecurrentTransfer {
                    from_account: self.from_account,
                    to_account: self.to_account,
                    amount,
                    memo: self.memo.unwrap_or_default(),
                    recurrence: self.recurrence.unwrap_or(DEFAULT_RECURRENCE_HOURS),
                    executions: self.executions.unwrap_or(DEFAULT_EXECUTIONS),
                    extensions: build_extensions(self.pair_id),
                },
            )),
        }])
    }
}

/// Builder for the removal flavour of recurrent transfer (zero-amount HIVE).
/// Mirrors `RecurrentTransferRemovalOperation` from TS.
#[derive(Debug, Clone)]
pub struct RecurrentTransferRemovalOperation {
    pub from_account: AccountName,
    pub to_account: AccountName,
    pub pair_id: Option<u32>,
}

impl OperationBuilder for RecurrentTransferRemovalOperation {
    fn finalize(
        self,
        foundation: &dyn WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        // TS defaults to `{ ...ASSETS.HIVE, amount: "0" }` when the amount is
        // undefined; the symbol parser handles HIVE-zero cleanly.
        let amount = foundation.hive_satoshis(0)?;

        Ok(vec![proto::Operation {
            value: Some(proto::operation::Value::RecurrentTransferOperation(
                proto::RecurrentTransfer {
                    from_account: self.from_account,
                    to_account: self.to_account,
                    amount,
                    memo: String::new(),
                    recurrence: DEFAULT_RECURRENCE_HOURS,
                    executions: DEFAULT_EXECUTIONS,
                    extensions: build_extensions(self.pair_id),
                },
            )),
        }])
    }
}
